composer/algorithms/agc/agc.py

```diff
@@ -0,0 +1,161 @@
+    Runs on ``Event.AFTER_TRAIN_BATCH``.
+
+    Example:
+         .. testcode::
+
+            from composer.algorithms import AGC
+            from composer.trainer import Trainer
+            agc_algorithm = AGC()
+            trainer = Trainer(
+                model=model,
+                train_dataloader=train_dataloader,
+                eval_dataloader=eval_dataloader,
+                max_duration="1ep",
+                algorithms=[agc_algorithm],
+                optimizers=[optimizer]
+            )
+
+    Args:
+        clipping_threshold (float, optional): The largest acceptable ratio between grad
+            norms and parameter norms before clipping is done.
+        eps (float, optional): Minimum value that weight norms are clamped to.
+    """
+
+    def __init__(self, clipping_threshold: float = 0.01, eps: float = 1e-3):
+        self.clipping_threshold = clipping_threshold
+        self.eps = eps
+
+    def match(self, event: Event, state: State) -> bool:
+        """Run on ``Event.AFTER_TRAIN_BATCH``."""
+        return event == Event.AFTER_TRAIN_BATCH
+
+    def apply(self, event: Event, state: State, logger: Logger) -> Optional[int]:
+        """Freeze layers in the model."""
+        apply_agc(model=state.model, clipping_threshold=self.clipping_threshold, eps=self.eps)
+
+
+# Factored this to a function to enable easier testing.
+def _get_clipped_gradient_coeff(weights: torch.Tensor,
+                                grad: torch.Tensor,
+                                clipping_threshold: float = 0.01,
+                                eps: float = 1e-3):
+    """Clips all gradients in model based on ratio of gradient norms to parameter norms.
+
+    Gradients whose norms exceed weight_norm * clipping_threshold are scaled down by
+    (weight_norm / grad_norm) * clipping_threshold.
+
+    Args:
+        weights (torch.Tensor): Tensor of weights (parameters) from the model.
+        grad (torch.Tensor): Tensor of gradients of the loss with respect to the weights.
+        clipping_threshold (float, optional): The largest acceptable ratio between grad
+            norms and parameter norms before clipping is done.
+        eps (float, optional): Minimum value that weight norms are clamped to.
+
+    Return:
+        clipped_grad_coeff (torch.Tensor): Coefficient of same shape as grad_norm equal to
+            (weight_norm / grad_norm) * clipping_threshold for gradients whose norms
+            that exceed weight_norm * clipping_threshold and one otherwise.
+    """
+
+    # Compute and clamp grad and weight norms.
+    w_norm = _unitwise_norm(weights).clamp_(min=eps)
+    grad_norm = _unitwise_norm(grad).clamp_(min=1e-6)
+
+    # Gradients whose norms are greater than weight_norm * clipping_threhsold are
+    # scaled down by (weight_norm * clipping_threhsold) / grad_norm.
+    max_norm = w_norm * clipping_threshold
+    clipped_grad_coeff = torch.where(grad_norm > max_norm, max_norm / grad_norm, torch.ones_like(grad_norm))
```

Comment:
Bit paranoid here, but could we do this in-place?
```
max_norm = w_norm.mul_(clipping_threshold)
clipped_grad_coeff = max_norm.div_(grad_norm).clamp_(max=1.0)

Reply:
oooh clever way of doing it fully in place. nice! ok fixed in 
2065360